### PR TITLE
Adding schema based object folders in OE

### PR DESCRIPTION
**Pull request by aasimkhan30**
When schema-based OE is enabled:
image

When schema-based OE is disabled: (Existing OE design)
image

It will be controlled via settings in ADS:
image

### CODE REVIEW

**Review by erinstellato-ms — APPROVED** [review]
My only request is that after the user changes the setting, we automatically refresh OE to show the top level view (e.g. if connected to instance, it would show the Databases, Security and Server Objects folder, and if connected to a database it would show Schemas, Programmability, etc.).  We definitely need to document this option and think about how to make it easily discoverable.

**Comment by alanrenmsft** [comment]
@aasimkhan30 I think we don't need the schemas folder under database node.

**Comment by cheenamalhotra** [comment]
> 
image

I would recommend we rename it to something like: "Mssql > Object Explorer: Group by Schema"

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs [hunk 1]
```diff
@@ -681,65 +694,164 @@ internal partial class DatabaseChildFactory : SmoChildFactoryBase
...
+                    NodeValue = SR.SchemaHierarchy_Views,
+                    NodeType = "Folder",
+                    NodeTypeId = NodeTypes.Views,
+                    IsSystemObject = false,
+                    SortPriority = SmoTreeNode.NextSortPriority,
+                });
+			}
+			if (!WorkspaceService<SqlToolsSettings>.Instance.CurrentSettings.SqlTools.ObjectExplorer.SchemaBasedOE)
+			{
+                currentChildren.Add(new FolderNode {
```
Comment (cheenamalhotra) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs:729]: Is there a usecase where a `FolderNode`'s `NodeType` will not be `Folder`?
Feels like something that can be taken care by `FolderNode` constructor, and could be avoided when creating these objects.

Can be followed up later.

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs [hunk 2]
```diff
@@ -681,65 +694,164 @@ internal partial class DatabaseChildFactory : SmoChildFactoryBase
...
             });
+        }
+
+        internal override Type[] ChildQuerierTypes
+        {
+            get
+            {
+                return new Type[0];
+            }
+        }
```
Comment (cheenamalhotra) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs:785]:
```suggestion
        internal override Type[] ChildQuerierTypes => new Type[0];
```

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs [hunk 3]
```diff
@@ -681,65 +694,164 @@ internal partial class DatabaseChildFactory : SmoChildFactoryBase
...
+    {
+        public override IEnumerable<string> ApplicableParents() { return new[] { "ExpandableSchemas" }; }
+
+        internal override Type[] ChildQuerierTypes
+        {
+            get
+            {
+                return new [] { typeof(SqlSchemaQuerier), };
+            }
+        }
```
Comment (cheenamalhotra) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs:809]:
```suggestion
        internal override Type[] ChildQuerierTypes => new [] { typeof(SqlSchemaQuerier), };
```

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs [hunk 4]
```diff
@@ -1004,35 +1115,97 @@ internal partial class ProgrammabilityChildFactory : SmoChildFactoryBase
...
+			}
+        }
+
+        internal override Type[] ChildQuerierTypes { get {return null;} }
+
+
+        public override TreeNode CreateChild(TreeNode parent, object context)
+        {
+            return null;
+        }
```
Comment (cheenamalhotra) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs:1185]:
```suggestion
        public override TreeNode CreateChild(TreeNode parent, object context) => null;
```

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs [hunk 5]
```diff
@@ -681,65 +694,164 @@ internal partial class DatabaseChildFactory : SmoChildFactoryBase
...
+            InitializeChild(parent, child, context);
+            return child;
+        }
+    }
+
+    [Export(typeof(ChildFactory))]
+    [Shared]
+    internal partial class ExpandableSchemasChildFactory : SmoChildFactoryBase
+    {
+        public override IEnumerable<string> ApplicableParents() { return new[] { "ExpandableSchemas" }; }
```
Comment (cheenamalhotra) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs:801]:
```suggestion
        public override IEnumerable<string> ApplicableParents() => new[] { "ExpandableSchemas" };
```

**Comment by aasimkhan30** [comment]
@alanrenmsft, updated screenshot. Now Schemas are under Database.
image

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModel.tt [hunk 6]
```diff
@@ -82,7 +83,37 @@ namespace Microsoft.SqlTools.ServiceLayer.ObjectExplorer.SmoModel
...
+						WriteLine("Values = new List<object> {{ {0}.{1} }},", parentVar, filterField);
+                        if(!string.IsNullOrWhiteSpace(filterValidFor))
+                        {
+                            WriteLine("ValidFor = {0},", GetValidForFlags(filterValidFor));
+                        }
+                        else
+                        {
+                            WriteLine("ValidFor = ValidForFlag.All ");
+                        }
+						PopIndent();
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModel.tt:111]: Fix indentation

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModelDefinition.xml [hunk 7]
```diff
@@ -35,14 +35,37 @@
 
   <Node Name="SqlErrorMessage" Type="UserDefinedMessage" Parent="Server" />
 
-  <Node Name="SqlTable" Parent="Database" />
+  <Node Name="SqlTable">
+    <Parent>Database</Parent>
+    <Parent>Schema</Parent>
+    <NavigationPath Parent="Database" Field="Tables" />
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModelDefinition.xml:41]: Please update the schema definition at the top with the additional functionality you're adding here (and making it more clear in general wouldn't hurt either, right now it's a little light on details and seems to be incomplete.

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModel.tt [hunk 8]
```diff
@@ -82,7 +83,37 @@ namespace Microsoft.SqlTools.ServiceLayer.ObjectExplorer.SmoModel
 			string navigationPath = GetNavigationPath(nodeElement, nodeName, navPathElement);
 			string subField = GetNavPathAttribute(navPathElement, "SubField");
 			string fieldType = GetNavPathAttribute(navPathElement, "FieldType");
-
+			if(navPathElement != null){
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModel.tt:86]: Add comments explaining what this code is doing

**Comment by alanrenmsft** [comment]
@aasimkhan30 can you expand a schema folder and the top level `Programability` folder? I am asking because some objects in it is also owned by schema.

**Comment by alanrenmsft** [comment]
also, in ADS side, I would also add a context menu item at server and database level: `Toggle Group By Schema` to make it more discoverable.

File: src/Microsoft.SqlTools.ServiceLayer/SqlContext/ObjectExplorerSettings.cs [hunk 9]
```diff
@@ -28,5 +28,11 @@ public ObjectExplorerSettings()
         /// Number of seconds to wait before fail expand request with timeout error
         /// </summary>
         public int ExpandTimeout { get; set; }
+
+        /// <summary>
+        /// Moves Schema to the top level of OE and then move Tables, Views, Programmability and Synonyms under it. 
+        /// </summary>
+        /// <value></value>
+        public bool SchemaBasedOE { get; set; }
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/SqlContext/ObjectExplorerSettings.cs:36]: As Cheena suggested I think just "GroupBySchema" here would be clearer

**Comment by Charles-Gagnon** [comment]
> also, in ADS side, I would also add a context menu item at server and database level: `Toggle Group By Schema` to make it more discoverable.

Even better, this should be a button action on the servers view itself. I'd avoid adding yet another context menu item for a global setting like this.

**Comment by alanrenmsft** [comment]
> > also, in ADS side, I would also add a context menu item at server and database level: `Toggle Group By Schema` to make it more discoverable.
> 
> Even better, this should be a button action on the servers view itself. I'd avoid adding yet another context menu item for a global setting like this.

this is a MSSQL only thing, so might not be a good idea to add it to the servers view.

**Comment by Charles-Gagnon** [comment]
Hrm, good point. Maybe an inline item on the top-level node? Anyways we can discuss that in the ADS PR

**Comment by aasimkhan30** [comment]
@alanrenmsft Expanding programmability.
azuredatastudio_1owlLkkuHs

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/INodeFilter.cs [hunk 10]
```diff
@@ -53,5 +53,25 @@ public static string GetPropertyFilter(IEnumerable<INodeFilter> filters, Type ty
...
+        /// <summary>
+        /// Appends a list of property filters to an URN query string
+        /// </summary>
+        public static string AddPropertyFilterToFilterString(string filterString, IEnumerable<INodeFilter> filters, Type type, ValidForFlag validForFlag)
+        {
+            if(String.IsNullOrEmpty(filterString))
+            {
+                return GetPropertyFilter(filters, type, validForFlag);
+            }
+            foreach(var value in filters)
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/INodeFilter.cs:66]: I think you do something like `foreach(var value in filters ?? Enumerable.Empty<INodeFilter>())` to be compliant with nullable compiler warnings for this statement.

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/NodeTypes.cs [hunk 11]
```diff
@@ -1,147 +1,143 @@
-﻿//
+
+//
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 //
 
+// This file was generated by a T4 Template. Do not modify directly, instead update the SmoTreeNodesDefinition.xml file
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/NodeTypes.cs:7]: I assume we didn't lose any manual edits when regenerating?

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/NodeTypes.tt [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+        // Adding types that are not present anywhere. These types are referred in tests.
+        retElements.Add("Folder");
+
+        var retList = retElements.ToList();
+        retList.Sort();
+        return retList;
+    }
+
+    public static List<XmlElement> GetChildren(string xmlFile, string parentName)
+	{
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/NodeTypes.tt:79]: nit: extra indentation in the next few methods

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/NodeTypes.tt [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+		XmlElement nodeElement = GetNodeElement(xmlFile, parentName);
+		XmlDocument doc = new XmlDocument();
+		doc.Load(xmlFile);
+
+		List<XmlElement> retElements = new List<XmlElement>();
+		XmlNodeList nodeList = doc.SelectNodes(string.Format("/ServerExplorerTree/Node[@Name='{0}']/Child", parentName));
+		foreach (var item in nodeList)
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/NodeTypes.tt:86]: should probably do the Enumerable<>.Empty() whenever an enumerator could be null.

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoChildFactoryBase.cs [hunk 14]
```diff
@@ -29,9 +29,18 @@ public override IEnumerable<TreeNode> Expand(TreeNode parent, bool refresh, stri
 
             try
             {
-                OnExpandPopulateFoldersAndFilter(allChildren, parent, includeSystemObjects);
-                RemoveFoldersFromInvalidSqlServerVersions(allChildren, parent);
-                OnExpandPopulateNonFolders(allChildren, parent, refresh, name, cancellationToken);
+                if(this.PutFoldersAfterNodes){
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoChildFactoryBase.cs:32]: nit: `{` should be on own line and spacing after `if`

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModel.cs [hunk 15]
```diff
@@ -356,6 +357,27 @@ public override  IEnumerable<SqlSmoObject> Query(SmoQueryContext context, string
...
+                    Values = new List<object> { parentSchema.Name },
+                    ValidFor = ValidForFlag.All 
+                });
+                filter = INodeFilter.AddPropertyFilterToFilterString(filter, filters, this.GetType(), context.ValidFor);
+                var retValue = parentSchema.Parent.Tables;
+                if (retValue != null)
+                {
+                    retValue.ClearAndInitialize(filter, extraProperties);
+                    var ret = new SmoCollectionWrapper<Table>(retValue).Where(c => PassesFinalFilters(parentSchema, c));
+                    Logger.Verbose("End query Table");
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModel.cs:377]: this log message seems to not have enough context to be useful.  Maybe at least include something about SmoQueryModel or ObjectExplorer, or doesn't the Logger automatically capture some context.

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModel.cs [hunk 16]
```diff
@@ -882,6 +1030,27 @@ public override  IEnumerable<SqlSmoObject> Query(SmoQueryContext context, string
                     return ret;
                 }
             }
+            Schema parentSchema = context.Parent as Schema;
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModel.cs:1033]: are all these duplicated snippets code-gened?  If not if looks like it could be a template function pretty simply.

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs [hunk 17]
```diff
@@ -677,73 +675,132 @@ public override TreeNode CreateChild(TreeNode parent, object context)
...
-            });
-            currentChildren.Add(new FolderNode {
-                NodeValue = SR.SchemaHierarchy_Synonyms,
-                NodeType = "Folder",
-                NodeTypeId = NodeTypes.Synonyms,
-                IsSystemObject = false,
-                ValidFor = ValidForFlag.AllOnPrem|ValidForFlag.AzureV12,
-                SortPriority = SmoTreeNode.NextSortPriority,
-            });
+			if (!WorkspaceService<SqlToolsSettings>.Instance.CurrentSettings.SqlTools.ObjectExplorer.GroupBySchema)
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.cs:683]: format document?

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.tt [hunk 18]
```diff
@@ -71,12 +74,13 @@ namespace Microsoft.SqlTools.ServiceLayer.ObjectExplorer.SmoModel
 	foreach (var TreeNode in allTreeNodes)
 	{
 		var name = TreeNode.GetAttribute("Name");
+		var type = !String.IsNullOrWhiteSpace(TreeNode.GetAttribute("Type")) ? TreeNode.GetAttribute("Type") : TreeNode.GetAttribute("Name");
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.tt:77]: `string.IsNullOrWhitespace`

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.tt [hunk 19]
```diff
@@ -23,13 +23,16 @@ using System.Collections.Generic;
...
+using Microsoft.SqlTools.ServiceLayer.SqlContext;
+using Microsoft.SqlTools.ServiceLayer.Workspace;
 
 namespace Microsoft.SqlTools.ServiceLayer.ObjectExplorer.SmoModel
 {
 
 <#
 	var directory = Path.GetDirectoryName(Host.TemplateFile);
 	string xmlFile = Path.Combine(directory, "SmoTreeNodesDefinition.xml");
+	var indent = "    ";
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoTreeNodes.tt:35]: if codegen is causing a lot of the messed up formatting in generated files then I won't spending too much time tuning, as it doesn't actually impact readability all that much.

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/ObjectExplorer/NodeFilterTests.cs [hunk 20]
```diff
@@ -191,5 +191,30 @@ public void GetPropertyFilterWithMixedFilters()
             string expectedTableQuerierFilters = "[(@IsSystemObject = 1) and ((@IsSystemObject = 1))]";
             Assert.That(invalidQuerierType, Is.EqualTo(expectedTableQuerierFilters), "GetPropertyFilter did not construct the URN filter string as expected when excluding filters that don't match the querier type.");
         }
+        [Test]
+        public void TestAddPropertyFiltersToExistingURNs()
```
Comment (kburtram) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/ObjectExplorer/NodeFilterTests.cs:195]: is this enough tests?

**Review by kburtram — APPROVED** [review]
Looks good overall.  There is a lot of changes and hard to know if there are side-effects, so we'll need to mitigate with testing.  Also, you may try to add some more tests given the amount of code changing.

**Comment by kburtram** [comment]
@aasimkhan30 can you verify the icons look good on dark themes?  And are correct in high-contrast?

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/INodeFilter.cs [hunk 21]
```diff
@@ -53,5 +53,25 @@ public static string GetPropertyFilter(IEnumerable<INodeFilter> filters, Type ty
...
+            {
+                var filter = value.ToPropertyFilterString(type, validForFlag);
+                if(filter != string.Empty)
+                {
+                    filterString = filterString.Remove(filterString.Length - 1, 1) + $" and {filter}" + "]";
```
Comment (lewis-sanchez) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/INodeFilter.cs:71]: What's the last square bracket for?

You could also get rid of the extra append operation by just adding the square bracket to the end of the interpolated string.

`filterString = filterString.Remove(filterString.Length - 1, 1) + $" and {filter}]"

**Review by lewis-sanchez — APPROVED** [review]

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/INodeFilter.cs (hunk 21 above)
Comment (aasimkhan30) [reply src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/INodeFilter.cs:71]: That's the last format for URN https://learn.microsoft.com/en-us/sql/powershell/query-expressions-and-uniform-resource-names?view=sql-server-ver16

**Comment by aasimkhan30** [comment]
> @aasimkhan30 can you verify the icons look good on dark themes? And are correct in high-contrast?

@kburtram, will follow up with those details in ADS PR.

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/NodeTypes.cs (hunk 11 above)
Comment (aasimkhan30) [reply src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/NodeTypes.cs:7]: I did do a check and have manually added those types back

File: src/Microsoft.SqlTools.ServiceLayer/Workspace/WorkspaceService.cs [hunk 22]
```diff
@@ -329,6 +329,7 @@ public void RegisterTextDocOpenCallback(TextDocOpenCallback task)
             {
                 Logger.Write(TraceEventType.Verbose, "HandleDidChangeConfigurationNotification");
 
+                this.CurrentSettings = configChangeParams.Settings;
```
Comment (alanrenmsft) [inline src/Microsoft.SqlTools.ServiceLayer/Workspace/WorkspaceService.cs:332]: looks like you are fixing a bug here, how does other services work without this value being up to date?

Comment (aasimkhan30) [reply src/Microsoft.SqlTools.ServiceLayer/Workspace/WorkspaceService.cs:332]: Most services that are using configs have registered a handler and are just using that.

**Review by alanrenmsft — APPROVED** [review]
overall looks good. a few more things:
1. please attach a screenshot for the scenario of  connecting to a database directly
2. profile all the expand operations and make sure a single operation are not causing multiple queries. (1 -> n)

**Comment by aasimkhan30** [comment]
@alanrenmsft DB tree expanded
image

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/NodeTypes.cs [hunk 23]
```diff
@@ -1,147 +1,143 @@
-﻿//
+
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/Nodes/NodeTypes.cs:1]: There generally shouldn't be a blank line before the copyright, remove line 12 from the template

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModelDefinition.xml (hunk 7 above)
Comment (Charles-Gagnon) [reply src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/SmoQueryModelDefinition.xml:41]: Ping @aasimkhan30

File: src/Microsoft.SqlTools.ServiceLayer/SqlContext/ObjectExplorerSettings.cs [hunk 24]
```diff
@@ -28,5 +28,11 @@ public ObjectExplorerSettings()
         /// Number of seconds to wait before fail expand request with timeout error
         /// </summary>
         public int ExpandTimeout { get; set; }
+
+        /// <summary>
+        /// Moves Schema to the top level of OE and then move Tables, Views, Programmability and Synonyms under it.
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/SqlContext/ObjectExplorerSettings.cs:33]:
```suggestion
        /// Moves Schema to the top level of OE and then move schema-bound nodes under it.
```

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by aasimkhan30** [comment]
@alanrenmsft  DB tree expanded
image

As for profiler:
TLDR: The schema-based query takes a little more time around 10-20% more because it has added a URN filter for schema in the query.

Here is an example of table folder expanded:
New:
azuredatastudio_dUZJ173tUo
Query executed:
```
exec sp_executesql N'SELECT
SCHEMA_NAME(tbl.schema_id) AS [Schema],
tbl.name AS [Name],
tbl.object_id AS [ID],
tbl.is_filetable AS [IsFileTable],
CAST(CASE WHEN (tbl.temporal_type = 2) OR (tbl.ledger_type = 2) THEN 1 ELSE 0 END AS bit) AS [IsSystemVersioned],
tbl.temporal_type AS [TemporalType],
CAST(ISNULL(tbl.ledger_type, 0) AS int) AS [LedgerType],
CAST(tbl.is_external AS bit) AS [IsExternal],
CAST(tbl.is_edge AS bit) AS [IsEdge],
CAST(tbl.is_node AS bit) AS [IsNode]
FROM
sys.tables AS tbl
WHERE
((CAST(
 case 
    when tbl.is_ms_shipped = 1 then 1
    when (
        select 
            major_id 
        from 
            sys.extended_properties 
        where 
            major_id = tbl.object_id and 
            minor_id = 0 and 
            class = 1 and 
            name = N''microsoft_database_tools_support'') 
        is not null then 1
    else 0
end          
             AS bit)=@_msparam_0) and (tbl.temporal_type=@_msparam_1 or tbl.temporal_type=@_msparam_2) and (CAST(ISNULL(tbl.ledger_type, 0) AS int)=@_msparam_3 or CAST(ISNULL(tbl.ledger_type, 0) AS int)=@_msparam_4 or CAST(ISNULL(tbl.ledger_type, 0) AS int)=@_msparam_5) and (CAST(ISNULL(tbl.is_dropped_ledger_table, 0) AS bit)=@_msparam_6) and (SCHEMA_NAME(tbl.schema_id)=@_msparam_7))
ORDER BY
[Schema] ASC,[Name] ASC',N'@_msparam_0 nvarchar(4000),@_msparam_1 nvarchar(4000),@_msparam_2 nvarchar(4000),@_msparam_3 nvarchar(4000),@_msparam_4 nvarchar(4000),@_msparam_5 nvarchar(4000),@_msparam_6 nvarchar(4000),@_msparam_7 nvarchar(4000)',@_msparam_0=N'0',@_msparam_1=N'0',@_msparam_2=N'2',@_msparam_3=N'0',@_msparam_4=N'3',@_msparam_5=N'2',@_msparam_6=N'0',@_msparam_7=N'db_accessadmin'
```

Old:
azuredatastudio-insiders_mOnTBa5vlb
```
exec sp_executesql N'SELECT
SCHEMA_NAME(tbl.schema_id) AS [Schema],
tbl.name AS [Name],
tbl.object_id AS [ID],
tbl.is_filetable AS [IsFileTable],
CAST(CASE WHEN (tbl.temporal_type = 2) OR (tbl.ledger_type = 2) THEN 1 ELSE 0 END AS bit) AS [IsSystemVersioned],
tbl.temporal_type AS [TemporalType],
CAST(ISNULL(tbl.ledger_type, 0) AS int) AS [LedgerType],
CAST(tbl.is_external AS bit) AS [IsExternal],
CAST(tbl.is_edge AS bit) AS [IsEdge],
CAST(tbl.is_node AS bit) AS [IsNode]
FROM
sys.tables AS tbl
WHERE
((CAST(
 case 
    when tbl.is_ms_shipped = 1 then 1
    when (
        select 
            major_id 
        from 
            sys.extended_properties 
        where 
            major_id = tbl.object_id and 
            minor_id = 0 and 
            class = 1 and 
            name = N''microsoft_database_tools_support'') 
        is not null then 1
    else 0
end          
             AS bit)=@_msparam_0) and (tbl.temporal_type=@_msparam_1 or tbl.temporal_type=@_msparam_2) and (CAST(ISNULL(tbl.ledger_type, 0) AS int)=@_msparam_3 or CAST(ISNULL(tbl.ledger_type, 0) AS int)=@_msparam_4 or CAST(ISNULL(tbl.ledger_type, 0) AS int)=@_msparam_5) and (CAST(ISNULL(tbl.is_dropped_ledger_table, 0) AS bit)=@_msparam_6))
ORDER BY
[Schema] ASC,[Name] ASC',N'@_msparam_0 nvarchar(4000),@_msparam_1 nvarchar(4000),@_msparam_2 nvarchar(4000),@_msparam_3 nvarchar(4000),@_msparam_4 nvarchar(4000),@_msparam_5 nvarchar(4000),@_msparam_6 nvarchar(4000)',@_msparam_0=N'0',@_msparam_1=N'0',@_msparam_2=N'2',@_msparam_3=N'0',@_msparam_4=N'3',@_msparam_5=N'2',@_msparam_6=N'0'
```

**Comment by alanrenmsft** [comment]
@aasimkhan30 make sure you also profile the expand action for: Database, Schema, Views, Programability.

**Comment by aasimkhan30** [comment]
@alanrenmsft, previously there were no queries executed for database expansion as it was only a list of folders:
image

Now, we are executing queries and it takes some time:

image

**Review by kburtram — APPROVED** [review]
latest changes look good to me
